ArchiSteamFarm/Core/Utilities.cs

```diff
@@ -322,37 +347,105 @@ public static class Utilities {
 		return (result.Score < 4, suggestions is { Count: > 0 } ? string.Join(' ', suggestions.Where(static suggestion => suggestion.Length > 0)) : null);
 	}
 
-	internal static bool UpdateFromArchive(ZipArchive zipArchive, string targetDirectory) {
+	internal static async Task<bool> UpdateFromArchive(ZipArchive zipArchive, string targetDirectory) {
 		ArgumentNullException.ThrowIfNull(zipArchive);
 		ArgumentException.ThrowIfNullOrEmpty(targetDirectory);
 
-		// Firstly we'll move all our existing files to a backup directory
-		string backupDirectory = Path.Combine(targetDirectory, SharedInfo.UpdateDirectory);
+		try {
+			// Firstly, ensure once again our directories are purged and ready to work with
+			foreach (string directory in Directory.EnumerateDirectories(targetDirectory, $"{SharedInfo.UpdateDirectoryNewPrefix}*")) {
+				ASF.ArchiLogger.LogGenericInfo(Strings.UpdateCleanup);
+
+				Directory.Delete(directory, true);
+
+				ASF.ArchiLogger.LogGenericInfo(Strings.Done);
+			}
+
+			foreach (string directory in Directory.EnumerateDirectories(targetDirectory, $"{SharedInfo.UpdateDirectoryOldPrefix}*")) {
+				ASF.ArchiLogger.LogGenericInfo(Strings.UpdateCleanup);
+
+				await DeletePotentiallyUsedDirectory(directory).ConfigureAwait(false);
+
+				ASF.ArchiLogger.LogGenericInfo(Strings.Done);
+			}
+
+			string oldBackupDirectory = Path.Combine(targetDirectory, SharedInfo.UpdateDirectoryOld);
+
+			if (Directory.Exists(oldBackupDirectory)) {
+				ASF.ArchiLogger.LogGenericInfo(Strings.UpdateCleanup);
+
+				await DeletePotentiallyUsedDirectory(oldBackupDirectory).ConfigureAwait(false);
+
+				ASF.ArchiLogger.LogGenericInfo(Strings.Done);
+			}
+		} catch (Exception e) {
+			ASF.ArchiLogger.LogGenericException(e);
+
+			return false;
+		}
+
+		// Create temporary directories, we append current unix timestamp to ensure that our folders are unique, which helps to avoid delete-create race conditions for the same path
+		ulong unixTime = GetUnixTime();
+
+		string backupDirectoryName = $"{SharedInfo.UpdateDirectoryOldPrefix}{unixTime}";
+		string updateDirectoryName = $"{SharedInfo.UpdateDirectoryNewPrefix}{unixTime}";
+
+		string backupDirectory = Path.Combine(targetDirectory, backupDirectoryName);
+		string updateDirectory = Path.Combine(targetDirectory, updateDirectoryName);
+
+		// Now enumerate over files in the zip archive and extract them to entirely new location, this decreases chance of corruptions if user kills the process during this stage
+		Directory.CreateDirectory(updateDirectory);
+
+		foreach (ZipArchiveEntry zipFile in zipArchive.Entries) {
+			switch (zipFile.Name) {
+				case null:
+				case "":
```

Comment:
1. Can `zipFile.Name` actually be null? Docs suggest this is a non-null value
2. If these cases are something that should never happen, ASF should probably throw an exception/return error since it would be an invalid archive